File: backend/app/schemas/product.py
from typing import List, Optional

from pydantic import BaseModel


class ProductCreate(BaseModel):
    p_id: str
    name: str
    price: float

    colour: str
    brand: str
    category: str

    img: str

    ratingCount: int = 0
    avg_rating: float = 0

    description: str

    p_attributes: List[str]


class ProductUpdate(BaseModel):
    name: Optional[str] = None
    price: Optional[float] = None

    colour: Optional[str] = None
    brand: Optional[str] = None
    category: Optional[str] = None

    img: Optional[str] = None

    ratingCount: Optional[int] = None
    avg_rating: Optional[float] = None

    description: Optional[str] = None

    p_attributes: Optional[List[str]] = None

    status: Optional[str] = None

    qualityScore: Optional[float] = None

File: frontend/types/product.ts
import type { ProductStatus } from '@/lib/constants/product-status';

/** Raw column shape used by the Myntra Fashion Product Dataset. */
export interface MyntraProductDatasetRow {
  p_id: string | number;
  name: string;
  price: number;
  colour: string;
  brand: string;
  img: string;
  ratingCount: number;
  avg_rating: number;
  description: string;
  p_attributes: ProductAttributes;
}

export type ProductAttributeValue = string | string[];
export type ProductAttributes = Record<string, ProductAttributeValue>;

/** Normalized product model used throughout the application. */
export interface Product {
  id: string;
  name: string;
  price: number;
  colour: string;
  brand: string;
  img: string;
  ratingCount: number;
  avgRating: number;
  description: string;
  attributes: ProductAttributes;
  status: ProductStatus;
  qualityScore: number;
  createdAt: string;
  updatedAt: string;
}

export type ProductInput = Omit<Product, 'id' | 'createdAt' | 'updatedAt'>;
